### PR TITLE
Optimize and fix some bugs for autoupload 

### PR DESCRIPTION
**Pull request by JonasMayerDev**
## Fixes
- Makes Autoupload file scan faster by checking for changes since last file system scan and not query database for last known change timestamp (saves the database query) 
- Fixes logic mistake to start foreground service when doing big scan and not for content observer events
- Fixes start of infinite ContentObservers 

### To test performance improvements
- Select a big custom folder for autoupload
- wait till all files are scheduled for upload (activate pause to not upload all the files) 
- In uploads, tab swipe down to restart file sync job
- Check how long the "preparing for upload" notification stays there
- Compare to version before https://github.com/nextcloud/android/pull/12719 since the foreground notification was broken with it

### To test if autoupload still working
- Add document (not video or image) to one of the autoupload folder 
- Start file sync worker by swiping down in uploads tab
- Wait till "prepare for upload" notification is gone
- Check if file was scheduled for upload

### Maybe also test
- Upgrade (Database changes)
- ContentObserver changed file detection by adding image to auto upload folder (should add to upload within 30s)

- [X] Tests written, or not not needed

### CODE REVIEW

File: app/src/main/java/com/owncloud/android/datamodel/SyncedFolder.java [hunk 1]
```diff
@@ -271,4 +277,8 @@ public void setExcludeHidden(boolean excludeHidden) {
     public boolean containsFile(String filePath){
         return filePath.contains(localPath);
     }
+
+    public long getLastScanTimestampMs() { return lastScanTimestampMs; }
```
Comment (alperozturk96) [inline app/src/main/java/com/owncloud/android/datamodel/SyncedFolder.java:281]: No need to give Ms suffix timestamp already explain itself

File: app/src/main/java/com/owncloud/android/utils/FilesSyncHelper.java [hunk 2]
```diff
@@ -63,60 +116,38 @@ private static void insertAllDBEntriesForSyncedFolder(SyncedFolder syncedFolder)
...
-                        public FileVisitResult visitFileFailed(Path file, IOException exc) {
-                            return FileVisitResult.CONTINUE;
-                        }
-                    });
-                } catch (IOException e) {
-                    Log_OC.e(TAG, "Something went wrong while indexing files for auto upload", e);
-                }
+                    long startTime = System.nanoTime();
+                    FilesSyncHelper.insertCustomFolderIntoDB(path, syncedFolder, filesystemDataProvider, lastCheck, thisCheck);
+                    Log_OC.d(TAG,"FILESYNC FINISHED LONG CHECK FOLDER "+path+" "+(System.nanoTime() - startTime));
```
Comment (alperozturk96) [inline app/src/main/java/com/owncloud/android/utils/FilesSyncHelper.java:142]: I prefer that the characters of general log messages are not capitalized.

File: app/src/main/java/com/owncloud/android/utils/FilesSyncHelper.java [hunk 3]
```diff
@@ -160,7 +192,7 @@ private static String getFileFromURI(String uri){
         return filePath;
     }
 
-    private static void insertContentIntoDB(Uri uri, SyncedFolder syncedFolder) {
+    private static void insertContentIntoDB(Uri uri, SyncedFolder syncedFolder, long lastCheckMs, long thisCheckMs) {
```
Comment (alperozturk96) [inline app/src/main/java/com/owncloud/android/utils/FilesSyncHelper.java:195]: Ms suffix? maybe lastCheckedTimestamp

**Review by alperozturk96 — APPROVED** [review]

**Review by ZetaTom — APPROVED** [review]
Ran all tests successfully; passed with flying colours. The changes I suggested are minor nitpicks.

- Upgraded from current master version to this branch
  - kept auto upload folders
  - kept progress info
  - auto upload worked as expected
- Uploaded large folder (4000 files)
- Uploaded several nested folders
- Much faster than the version master
- Actually fixes some weirdness in the *Uploads* view

File: app/src/main/java/com/owncloud/android/datamodel/SyncedFolder.java (hunk 1 above)
Comment (JonasMayerDev) [reply app/src/main/java/com/owncloud/android/datamodel/SyncedFolder.java:281]: Last modified timestamp from content resolver is in sec. so it is not obv.

File: app/src/main/java/com/owncloud/android/utils/FilesSyncHelper.java [hunk 4]
```diff
@@ -63,60 +116,38 @@ private static void insertAllDBEntriesForSyncedFolder(SyncedFolder syncedFolder)
...
-                            return FileVisitResult.CONTINUE;
-                        }
-                    });
-                } catch (IOException e) {
-                    Log_OC.e(TAG, "Something went wrong while indexing files for auto upload", e);
-                }
+                    long startTime = System.nanoTime();
+                    FilesSyncHelper.insertCustomFolderIntoDB(path, syncedFolder, filesystemDataProvider, lastCheck, thisCheck);
+                    Log_OC.d(TAG,"FILESYNC FINISHED LONG CHECK FOLDER "+path+" "+(System.nanoTime() - startTime));
+
```
[the remark this answers is not in the document]
Comment (JonasMayerDev) [reply app/src/main/java/com/owncloud/android/utils/FilesSyncHelper.java]: I think it makes sense for debug purposes to keep the time it took to scan a folder

**Pull request merged**